src/boot/efi/util.c

```diff
@@ -769,6 +770,111 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         return EFI_SUCCESS;
 }
 
+#define QEMU_KERNEL_LOADER_FS_MEDIA_GUID \
+        { 0x1428f772, 0xb64a, 0x441e, {0xb8, 0xc3, 0x9e, 0xbd, 0xd7, 0xf8, 0x93, 0xc7 }}
+
+#define QEMU_BOOT_ORDER_GUID \
+        { 0x668f4529, 0x63d0, 0x4bb5, {0xb6, 0x5d, 0x6f, 0xbb, 0x9d, 0x36, 0xa4, 0x4a }}
+
+/* detect direct boot */
+bool is_direct_boot(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        /* 'qemu -kernel systemd-bootx64.efi' */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_VENDOR_DP &&
+            memcmp(&dp->Guid, &(EFI_GUID)QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) == 0)
+                return true;
+
+        /* loaded from firmware volume (sd-boot added to ovmf) */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_PIWG_FW_VOL_DP)
+                return true;
+
+        return false;
+}
+
+/* return the length of the device path in bytes, excluding the end tag */
+static size_t device_path_length(const EFI_DEVICE_PATH *dp) {
+        assert(dp);
+
+        const EFI_DEVICE_PATH *end_node = dp;
+        while (!IsDevicePathEnd(end_node))
+                end_node = NextDevicePathNode(end_node);
+        return (uint8_t*)end_node - (uint8_t*)dp;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *ret_qemu_dev, EFI_FILE **ret_qemu_dir) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles;
+        EFI_STATUS err;
+
+        assert(ret_qemu_dev);
+        assert(ret_qemu_dir);
+
+        /* ovmf doesn't connect devices for direct kernel boot */
+        reconnect();
+
+        /* find all file system handles */
+        err = BS->LocateHandleBuffer(ByProtocol, &FileSystemProtocol, NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return err;
+
```

Comment:
hmm, is the QemuBootOrder stuff and the other qemu apis here, officially documented and considered stable apis? can you add references to relevant doc URls here?

Reply:
QemuBootOrder is brand new.

Direct kernel boot detects the pseudo filesystem used by ovmf to allow the linux efi stub load the initrd.  Not an official API but unchanged since years.  Not fully sure where we stand in terms of non-qemu hypervisors.  I know some (freebsd bhyve for example) mimic qemu fw_cfg, so this could work as-is if the host-side support for loading kernel+initrd via fw_cfg is there.

Firmware volume detection is standard edk2 stuff and not qemu specific.  If you feel advantagous and want add stuff to your physical machine firmware flash this should work too.  If you want add sd-boot to your ovmf builds look at my experimental here: https://github.com/kraxel/edk2/commits/sd-boot